floyd/cli/utils.py

```diff
@@ -55,28 +60,139 @@ def get_data_id(data_str):
+        else:
+            # foo
+            name = name_parts[0]
     else:
-        name_parts = data_name.split('/')
-        if len(name_parts) <= 3:
-            name_parts.insert(1, 'datasets')
-            data_name = '/'.join(name_parts)
-        return data_name
+        return raw_name
+
+    # If no number/version is found, query the API for the most recent version
+    if number is None:
+        name_from_api = get_lateset_dataset_version(username, name)
+        if not name_from_api:
+            raise FloydException("Could not resolve %s. Make sure the project exists and has jobs." % raw_name)
+        return name_from_api
+
+    return '/'.join([username, 'datasets', name, number])
+
+
+def normalize_job_name(raw_job_name, default_username=None, default_project_name=None):
+    name_parts = raw_job_name.split('/')
+
+    username = default_username or current_username()
+    project_name = default_project_name or current_experiment_name()
+    number = None  # current job number
+
+    # When nothing is passed, use all the defaults
+    if not raw_job_name:
+        pass
+    elif len(name_parts) == 4:
+        # mckay/projects/foo/1
+        username, _, project_name, number = name_parts
+    elif len(name_parts) == 3:
+
+        if name_parts[2].isdigit():
+            # mckay/foo/1
+            username, project_name, number = name_parts
+        else:
+            # mckay/projects/foo
+            username, _, project_name = name_parts
+    elif len(name_parts) == 2:
+        if name_parts[1].isdigit():
+            # foo/1
+            project_name, number = name_parts
+        else:
+            # mckay/foo
+            username, project_name = name_parts
+    elif len(name_parts) == 1:
+        if name_parts[0].isdigit():
+            # 1
+            number = name_parts[0]
+        else:
+            # foo
+
+            project_name = name_parts[0]
+    else:
+        return raw_job_name
 
+    # If no number is found, query the API for the most recent job number
+    if number is None:
+        job_name_from_api = get_latest_job_name(username, project_name)
+        if not job_name_from_api:
+            raise FloydException("Could not resolve %s. Make sure the project exists and has jobs." % raw_job_name)
+        return job_name_from_api
 
-def normalize_job_name(job_name):
-    job_name_parts = job_name.split('/')
-    if len(job_name_parts) <= 3:
-        job_name_parts.insert(1, 'projects')
-        job_name = '/'.join(job_name_parts)
-    return job_name
+    return '/'.join([username, 'projects', project_name, number])
 
 
 def get_cli_version():
     return pkg_resources.require("floyd-cli")[0].version
+
+
+def current_username():
+    access_token = AuthConfigManager.get_access_token()
+    return access_token.username
+
+
+def current_experiment_name():
+    experiment_config = ExperimentConfigManager.get_config()
+    return experiment_config.name
+
+
+def get_latest_job_name(username, project_name):
```

Comment:
Why is the behavior of the job different from the dataset? i.e., why get the latest name for job but latest version for dataset?

Reply:
Because project doesn't have the latest version api, so it's using the summary api. We can add the same api for project to keep it consistent.